diff --git a/build.sh b/build.sh
index dede74d..621588e 100755
--- a/build.sh
+++ b/build.sh
@@ -1,4 +1,5 @@
 #!/bin/bash
+set -o errexit
 
 source ./config.sh
 
@@ -19,7 +20,7 @@ echo " "
 echo "Creating Alchemist executable:"
 echo " "
 cd ./build/$SYSTEM/
-make
+make -j8
 cd ../..
 echo " "
 echo $LINE
diff --git a/build/MacOS/Makefile b/build/MacOS/Makefile
index 571f4c2..3b4078c 100755
--- a/build/MacOS/Makefile
+++ b/build/MacOS/Makefile
@@ -10,12 +10,12 @@ CXXFLAGS += "-I$(ELEMENTAL_PATH)/include"
 CXXFLAGS += "-I$(SPDLOG_PATH)/include"
 CXXFLAGS += "-I$(EIGEN3_PATH)/include"
 CXXFLAGS += "-I$(ARPACK_PATH)/include"
+CXXFLAGS += -I/root/asio/asio/include "-L$(ARPACK_PATH)/lib" "-Wl,-rpath,$(ARPACK_PATH)/lib"
 
 LDFLAGS += "-L$(EL_LIB)" "-Wl,-rpath,$(EL_LIB)" $(EL_LIBS)
 LDFLAGS += -Wl,-rpath,$(ARPACK_PATH)/lib -larpack
-LDFLAGS += -Wl,-rpath,$(SPDLOG_PATH)/lib -lspdlog
 # LDFLAGS += -lboost_system-mt -lboost_thread-mt
-LDFLAGS += -lmpi
+LDFLAGS += -lmpi -ldl
 
 OBJ_FILES = \
 	$(TARGET_PATH)/main.o \
diff --git a/src/main/GroupWorker.cpp b/src/main/GroupWorker.cpp
index 45f3d35..cefb9f4 100644
--- a/src/main/GroupWorker.cpp
+++ b/src/main/GroupWorker.cpp
@@ -18,7 +18,7 @@ GroupWorker::GroupWorker(Group_ID _group_ID, Worker_ID _worker_ID, io_context &
 			Server(_io_context, endpoint, _log), library(nullptr), group_ID(_group_ID), group(MPI_COMM_NULL), group_peers(MPI_COMM_NULL), worker_ID(_worker_ID),
 			next_session_ID(0), accept_connections(false)
 {
-	char buffer[12];
+	char buffer[13];
 	sprintf(buffer, "worker-%03d", worker_ID);
 
 	log = start_log(string(buffer), "[%Y-%m-%d %H:%M:%S.%e] [%n] [%l] %v");
diff --git a/src/main/utility/endian.hpp b/src/main/utility/endian.hpp
index 3f513fd..cfee179 100644
--- a/src/main/utility/endian.hpp
+++ b/src/main/utility/endian.hpp
@@ -20,7 +20,7 @@
   #define be64toh(x) OSSwapBigToHostInt64(x)
   #define le64toh(x) OSSwapLittleToHostInt64(x)
 
-#elif defined(__LINUX__)
+#elif defined(__linux__)
 
   #include <endian.h>
 
